### PR TITLE
Add AS7341 spectral color sensor

### PR DESCRIPTION
**Pull request by mrgnr**
# What does this implement/fix?

This adds support for the AS7341 sensor ([Adafruit](https://www.adafruit.com/product/4698)) using I2C.

References:
- [Datasheet](https://ams.com/documents/20143/36005/AS7341_DS000504_3-00.pdf/5eca1f59-46e2-6fc5-daf5-d71ad90c9b2b)
- [Adafruit_AS7341](https://github.com/adafruit/Adafruit_AS7341) (Implementation for Arduino)
- [SparkFun_AS7341X_Arduino_Library](https://github.com/sparkfun/SparkFun_AS7341X_Arduino_Library) (Implementation for Arduino)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2609

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

Screenshot from **Home Assistant**:
as7341-homeassistant

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
- platform: as7341
  gain: X8
  atime: 120
  astep: 99
  f1:
    name: "415nm"
  f2:
    name: "445nm"
  f3:
    name: "480nm"
  f4:
    name: "515nm"
  f5:
    name: "555nm"
  f6:
    name: "590nm"
  f7:
    name: "630nm"
  f8:
    name: "680nm"
  clear:
    name: "Clear"
  nir:
    name: "NIR"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/as7341/sensor.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def sensor_schema():
+    return sensor.sensor_schema(
+        unit_of_measurement=UNIT_COUNTS,
+        icon=ICON_BRIGHTNESS_5,
+        accuracy_decimals=0,
+        device_class=DEVICE_CLASS_ILLUMINANCE,
+        state_class=STATE_CLASS_MEASUREMENT,
+    )
```
Comment (jesserockz) [inline esphome/components/as7341/sensor.py]: No need to wrap in a function
```suggestion
SENSOR_SCHEMA = sensor.sensor_schema(
    unit_of_measurement=UNIT_COUNTS,
    icon=ICON_BRIGHTNESS_5,
    accuracy_decimals=0,
    device_class=DEVICE_CLASS_ILLUMINANCE,
    state_class=STATE_CLASS_MEASUREMENT,
)
```

File: esphome/components/as7341/sensor.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            cv.Optional(CONF_F1): sensor_schema(),
+            cv.Optional(CONF_F2): sensor_schema(),
+            cv.Optional(CONF_F3): sensor_schema(),
+            cv.Optional(CONF_F4): sensor_schema(),
+            cv.Optional(CONF_F5): sensor_schema(),
+            cv.Optional(CONF_F6): sensor_schema(),
+            cv.Optional(CONF_F7): sensor_schema(),
+            cv.Optional(CONF_F8): sensor_schema(),
+            cv.Optional(CONF_CLEAR): sensor_schema(),
+            cv.Optional(CONF_NIR): sensor_schema(),
```
Comment (jesserockz) [inline esphome/components/as7341/sensor.py]:
```suggestion
            cv.Optional(CONF_F1): SENSOR_SCHEMA,
            cv.Optional(CONF_F2): SENSOR_SCHEMA,
            cv.Optional(CONF_F3): SENSOR_SCHEMA,
            cv.Optional(CONF_F4): SENSOR_SCHEMA,
            cv.Optional(CONF_F5): SENSOR_SCHEMA,
            cv.Optional(CONF_F6): SENSOR_SCHEMA,
            cv.Optional(CONF_F7): SENSOR_SCHEMA,
            cv.Optional(CONF_F8): SENSOR_SCHEMA,
            cv.Optional(CONF_CLEAR): SENSOR_SCHEMA,
            cv.Optional(CONF_NIR): SENSOR_SCHEMA,
```

File: esphome/components/as7341/as7341.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  AS7341_ADC_CHANNEL_5,
+};
+
+using as7341_smux_cmd_t = enum {
+  AS7341_SMUX_CMD_ROM_RESET,  ///< ROM code initialization of SMUX
+  AS7341_SMUX_CMD_READ,       ///< Read SMUX configuration to RAM from SMUX chain
+  AS7341_SMUX_CMD_WRITE,      ///< Write SMUX configuration from RAM to SMUX chain
+};
+
+using as7341_gain_t = enum {
```
Comment (jesserockz) [inline esphome/components/as7341/as7341.h]: Should be using this format to match the rest of ESPHome codebase =)
Same for the above enums
```suggestion
enum AS7341Gain {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by mrgnr** [comment]
Thanks for the review @jesserockz, I've made the requested changes and tested the sensor locally.

**Comment by jupe** [comment]
Thanks for component. Could you give home-assistant view config example here as well?

File: esphome/components/as7341/as7341.cpp [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+  LOG_SENSOR("  ", "F7", this->f7_);
+  LOG_SENSOR("  ", "F8", this->f8_);
+  LOG_SENSOR("  ", "Clear", this->clear_);
+  LOG_SENSOR("  ", "NIR", this->nir_);
+}
+
+float AS7341Component::get_setup_priority() const { return setup_priority::DATA; }
+
+void AS7341Component::update() {
+  read_channels(this->channel_readings_);
```
Comment (jesserockz) [inline esphome/components/as7341/as7341.cpp]: Consistency is key =)
```suggestion
  this->read_channels(this->channel_readings_);
```

File: esphome/components/as7341/as7341.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+};
+
+class AS7341Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void setup() override;
+  void dump_config() override;
+  float get_setup_priority() const override;
+  void update() override;
+
+  void set_f1_sensor(sensor::Sensor *f1_sensor) { f1_ = f1_sensor; }
```
Comment (jesserockz) [inline esphome/components/as7341/as7341.h]:
```suggestion
  void set_f1_sensor(sensor::Sensor *f1_sensor) { this->f1_ = f1_sensor; }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**